### ISSUE TITLE
[Numbering] Page number display error

### ISSUE THREAD

**Issue by kands-code** [typst/typst#5166]
### Description

The page number is displayed incorrectly on the page, but it is the correct page number in the table of contents.

<details>

<summary> page </summary>

![image](https://github.com/user-attachments/assets/e9a95002-e140-4da0-a4fd-bc851f5f1d9c)

</details>

<details>

<summary> toc </summary>

![image](https://github.com/user-attachments/assets/529e58ad-6c22-43bd-b3ac-a6ffa61dd1d8)

</details>

<details>

<summary> the corresponding code is </summary>

```typst
// simple_book.typ
// -- snip --

// contents
set page(numbering: "I", number-align: bottom + center)
counter(page).update(1)
outline(
  title: if contents_title == none { [Contents] } else { contents_title },
  indent: 2em,
)
pagebreak()

// preface
if preface == none {
  set page(numbering: none)
  pagebreak()
} else {
  include preface
  pagebreak()
}

// -- snip --

// preface.typ
#set page(numbering: "I", number-align: top + right)
#counter(page).update(1)

= 序

test

#pagebreak()

test
```

</details>

All the codes are in: [examples.zip](https://github.com/user-attachments/files/17306822/examples.zip)

I think the problem is caused by the order of rendering, i.e. the page numbers are rendered first and then updated. But I don't know the exact cause.

> typst version: `typst 0.11.1 (50115102)`

### Operating system

Linux, macOS

### Typst version

- [X] I am using the latest version of Typst

**Comment by laurmaedje** [comment]
What would be very helpful is if you could try to minimize the example by removing everything that is not relevant to the problem and putting everything into one file. That makes it much easier to follow what you're trying to achieve and how it is going wrong.

You could also try compiling with typst 0.12.0-rc1 and see whether that fixes your issue.

**Comment by kands-code** [comment]
> What would be very helpful is if you could try to minimize the example by removing everything that is not relevant to the problem and putting everything into one file. That makes it much easier to follow what you're trying to achieve and how it is going wrong.

The minimum reproducible code for this issue:

```typst
#outline()

#set page(numbering: "I", number-align: bottom + center)
#counter(page).update(1)

= First

test

#pagebreak()

#set page(numbering: "A", number-align: top + center)
#counter(page).update(1)

= Second

test
```

The result of the rendering is as follows:

---

**Table of Contents**

![toc](https://github.com/user-attachments/assets/d76afd9a-1bbe-4e3e-a412-74741308badf)

---

**Page**

![page](https://github.com/user-attachments/assets/51b0b4ef-d2fc-4668-8c03-9188bcadc4a3)

**Comment by kands-code** [comment]
> You could also try compiling with typst 0.12.0-rc1 and see whether that fixes your issue.

![0.12.0-rc1](https://github.com/user-attachments/assets/d60f78b7-287f-4d9d-b9a3-76c95a1c9037)

It looks like this issue hasn't been solved.

**Comment by EpicEricEE** [comment]
From what I can tell and as described in https://github.com/typst/typst/issues/3835#issuecomment-2304659477, this is the intended behaviour, so having the counter update between the page break and the set rule seems to be the way to go.

**Comment by kands-code** [comment]
> so having the counter update between the page break and the set rule seems to be the way to go

But if I put the update before the set rule, then the page number style is wrong:

<details>

<summary> show example </summary>

```typst
#set page(paper: "cn-business-card", flipped: true, width: 6.4cm)

#outline()

#counter(page).update(1)
#set page(numbering: "I", number-align: bottom + center)

= First

test

#pagebreak()

#counter(page).update(1)
#set page(numbering: "A", number-align: top + center)

= Second

test
```

![intended?](https://github.com/user-attachments/assets/8ada6161-dc40-49e9-baa5-37e6dead6242)

</details>

> Similar to https://github.com/typst/typst/issues/2987. If you move the state update above the set rule, it works (though you need to update to 0 then). Also note that counter updates do not need context, only retrievals.

and if I update the counter to 0, the counter is also wrong:

<details>

<summary> show example </summary>

```typst
#set page(paper: "cn-business-card", flipped: true, width: 6.4cm)

#outline()

#counter(page).update(1)
#set page(numbering: "I", number-align: bottom + center)

= First

test

#pagebreak()

#counter(page).update(0)
#set page(numbering: "A", number-align: top + center)

= Second

test
```

![work?](https://github.com/user-attachments/assets/9e36d3d8-79b1-42bf-92ab-bec5507384a0)

</details>

and if I use the "1 / 1" numbering style, the total page number is also affected, which as you said in issue #3835 is "intended", but I think it must be a bug.

<details>

<summary> show example </summary>

```typst
#set page(paper: "cn-business-card", flipped: true, width: 3.2cm)

#outline()

#set page(numbering: "I / I", number-align: bottom + center)
#counter(page).update(1)

= First

test

#pagebreak()

another test

#pagebreak()

#set page(numbering: "A / A", number-align: top + center)
#counter(page).update(1)

= Second

test
```

![bug?](https://github.com/user-attachments/assets/01aa5d8e-c64e-4fb4-8778-91d948ab43e2)

</details>

or is Typst not able to do "update page numbers while updating page styles" yet?

**Comment by kands-code** [comment]
Perhaps the problem is related to the fact that the rendering of pages is not linear, or that the state of each page is not independent? Because if it's linear or independent, the content of the later pages shouldn't affect the content of the earlier pages. Or is this design also "intended"?

All of the above are my own guesses, because I'm not going to read the code, and you can't expect every user to be able to read your code.

And I think this issue might be related to issue #5171.

**Comment by laurmaedje** [comment]
> But if I put the update before the set rule, then the page number style is wrong:

This is fixed in v0.12.0-rc1. So in 0.11.1, there was indeed a [bug](https://github.com/typst/typst/issues/2987), but it's fixed since.

> And I think this issue might be related to issue https://github.com/typst/typst/issues/5171.

That issue is different. It's caused by RTL language.

> All of the above are my own guesses, because I'm not going to read the code, and you can't expect every user to be able to read your code.

Nobody asked you to read the code. Please stay friendly.

**Comment by kands-code** [comment]
Ok, I will close this issue as it has been fixed in the next release.
